docs/running_lbann.rst

```diff
@@ -1,96 +1,434 @@
+          :language: python
 
-====================
+============================================================
 Running LBANN
-====================
+============================================================
 
-The basic template for running LBANN is
+------------------------------------------------
+Anatomy of an LBANN experiment
+------------------------------------------------
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Parallelism
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+LBANN is run under the `MPI
+<https://en.wikipedia.org/wiki/Message_Passing_Interface>`_ paradigm,
+i.e. with multiple processes that communicate with message
+passing. These processes are subdivided into "trainers." Conceptually,
+a trainer owns parallel objects, like models and data readers, and
+generally operates independently of other trainers.
+
+Comments:
+
++ LBANN targets HPC systems with homogeneous compute nodes and GPU
+  accelerators, which motivates some simplifying assumptions:
+
+  - All trainers have the same number of processes.
+
+  - Each MPI process corresponds to one GPU.
+
++ Processors are block assigned to trainers based on MPI rank.
+
+  - In order to minimize the cost of intra-trainer communication, make
+    sure to map processes to the hardware and network
+    topologies. Typically, this just means choosing a sensible number
+    of processes per trainer, e.g. a multiple of the number of GPUs
+    per compute node.
+
++ Generally, increasing the number of processes per trainer will
+  accelerate computation but require more intra-trainer
+  communication. There is typically a sweet spot where run time is
+  minimized, but it is complicated and sensitive to the type of
+  computational operations, the amount of work, the hardware and
+  network properties, and the communication algorithms.
+
+  - Rule-of-thumb: Configure experiments so that the bulk of run time
+    is taken by compute-bound operations (e.g. convolution or matrix
+    multiplication) and so that each process has enough work to
+    achieve a large fraction of peak performance (e.g. by making the
+    mini-batch size sufficiently large).
+
++ Most HPC systems are managed with job schedulers like `Slurm
+  <https://slurm.schedmd.com/overview.html>`_. Typically, users can
+  not immediately access compute nodes but must request them from
+  login nodes. The login nodes can be accessed directly (e.g. via
+  :bash:`ssh`), but users are discouraged from doing heavy computation
+  on them.
+
+  - For debugging and quick testing, it's convenient to request an
+    interactive session (:bash:`salloc` or :bash:`sxterm` with Slurm).
+
+  - If you need to run multiple experiments or if experiments are not
+    time-sensitive, it's best to submit a batch job (:bash:`sbatch`
+    with Slurm).
+
+  - When running an experiment, make sure you know what scheduler
+    account to charge (used by the scheduler for billing and
+    determining priority) and what scheduler partition to run on
```

Comment:
Might be good to add the options (e.g. `-p` or `--partition` for partition) for scheduler account and partition.